### PR TITLE
Refactor duplication endpoint

### PR DESCRIPTION
**Pull request by anistouri**

### CODE REVIEW

File: src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java [hunk 1]
```diff
@@ -61,12 +61,12 @@ public void insertScriptContingencyList(UUID id, String content) {
         restTemplate.exchange(actionsServerBaseUri + path, HttpMethod.POST, httpEntity, Void.class);
     }
 
-    public void insertScriptContingencyList(UUID sourceListId, UUID id) {
+    public UUID duplicateScriptContingencyList(UUID scriptContingencyListsId) {
         String path = UriComponentsBuilder.fromPath(DELIMITER + ACTIONS_API_VERSION + "/script-contingency-lists")
-                .queryParam(HEADER_DUPLICATE_FROM, sourceListId)
-                .queryParam("id", id)
+                .queryParam(HEADER_DUPLICATE_FROM, scriptContingencyListsId)
+                .buildAndExpand(scriptContingencyListsId)
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java:67]: to remove ther isn't any {id} in the path to replace

File: src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java [hunk 2]
```diff
@@ -89,20 +89,20 @@ public void insertIdentifierContingencyList(UUID id, String content) {
         restTemplate.exchange(actionsServerBaseUri + path, HttpMethod.POST, httpEntity, Void.class);
     }
 
-    public void insertFormContingencyList(UUID sourceListId, UUID id) {
+    public UUID duplicateFormContingencyList(UUID formContingencyListsId) {
         String path = UriComponentsBuilder.fromPath(DELIMITER + ACTIONS_API_VERSION + "/form-contingency-lists")
-                .queryParam(HEADER_DUPLICATE_FROM, sourceListId)
-                .queryParam("id", id)
+                .queryParam(HEADER_DUPLICATE_FROM, formContingencyListsId)
+                .buildAndExpand(formContingencyListsId)
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java:95]: same

File: src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java [hunk 3]
```diff
@@ -89,20 +89,20 @@ public void insertIdentifierContingencyList(UUID id, String content) {
...
+        return restTemplate.exchange(actionsServerBaseUri + path, HttpMethod.POST, null, UUID.class).getBody();
     }
 
-    public void insertIdentifierContingencyList(UUID sourceListId, UUID id) {
+    public UUID duplicateIdentifierContingencyList(UUID identifierContingencyListsId) {
         String path = UriComponentsBuilder.fromPath(DELIMITER + ACTIONS_API_VERSION + "/identifier-contingency-lists")
-                .queryParam(HEADER_DUPLICATE_FROM, sourceListId)
-                .queryParam("id", id)
+                .queryParam(HEADER_DUPLICATE_FROM, identifierContingencyListsId)
+                .buildAndExpand(identifierContingencyListsId)
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java:103]: same

File: src/main/java/org/gridsuite/explore/server/services/CaseService.java [hunk 4]
```diff
@@ -78,9 +78,10 @@ UUID importCase(MultipartFile multipartFile) {
         return caseUuid;
     }
 
-    UUID duplicateCase(UUID sourceCaseUuid) {
+    UUID duplicateCase(UUID caseId) {
         String path = UriComponentsBuilder.fromPath(DELIMITER + CASE_SERVER_API_VERSION + "/cases")
-                .queryParam("duplicateFrom", sourceCaseUuid)
+                .queryParam("duplicateFrom", caseId)
+                .buildAndExpand(caseId)
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/CaseService.java:84]: to remove

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java [hunk 5]
```diff
@@ -285,4 +257,22 @@ public void createNetworkModifications(List<ElementAttributes> modificationAttri
             }
         });
     }
+
+    public void duplicateContingencyList(UUID contingencyListsId, UUID targetDirectoryId, String userId, ContingencyListType contingencyListType) {
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/ExploreService.java:261]: move code near `createScriptContingencyList` for readability

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java [hunk 6]
```diff
@@ -87,6 +87,25 @@ public ElementAttributes createElementWithNewName(ElementAttributes elementAttri
...
 
+    public ElementAttributes duplicateElement(UUID elementUuid, UUID newElementUuid, UUID targetDirectoryId, String userId) {
+        UriComponentsBuilder uri = UriComponentsBuilder
+                .fromPath(ELEMENTS_SERVER_ROOT_PATH)
+                .queryParam("duplicateFrom", elementUuid)
+                .queryParam("newElementUuid", newElementUuid);
+        if (targetDirectoryId != null) {
+            uri.queryParam("targetDirectoryId", targetDirectoryId);
+        }
+        String path = uri.buildAndExpand(elementUuid)
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/DirectoryService.java:98]: you want to buildAndExpand a property which is absent in the path again

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java [hunk 7]
```diff
@@ -87,6 +87,25 @@ public ElementAttributes createElementWithNewName(ElementAttributes elementAttri
...
+                .toUriString();
+        HttpHeaders headers = new HttpHeaders();
+        headers.add(HEADER_USER_ID, userId);
+        headers.setContentType(MediaType.APPLICATION_JSON);
+        HttpEntity<ElementAttributes> httpEntity = new HttpEntity<>(headers);
+        return restTemplate
+                .exchange(directoryServerBaseUri + path, HttpMethod.POST, httpEntity, ElementAttributes.class)
+                .getBody();
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/DirectoryService.java:106]:
```suggestion
        return restTemplate
                .exchange(directoryServerBaseUri + path, HttpMethod.POST, new HttpEntity<>(headers), ElementAttributes.class)
                .getBody();
```

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java [hunk 8]
```diff
@@ -285,4 +257,22 @@ public void createNetworkModifications(List<ElementAttributes> modificationAttri
...
+            case SCRIPT -> contingencyListService.duplicateScriptContingencyList(contingencyListsId);
+            case FORM -> contingencyListService.duplicateFormContingencyList(contingencyListsId);
+            case IDENTIFIERS -> contingencyListService.duplicateIdentifierContingencyList(contingencyListsId);
+        };
+        directoryService.duplicateElement(contingencyListsId, newId, targetDirectoryId, userId);
+    }
+
+    public void duplicateNetworkModifications(UUID sourceId, UUID parentDirectoryUuid, String userId) {
+        // create duplicated modification
+        Map<UUID, UUID> newModificationsUuids = networkModificationService.createModifications(List.of(sourceId));
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/ExploreService.java:272]:
```suggestion
        Map<UUID, UUID> newModificationsUuids = networkModificationService.duplicateModifications(List.of(sourceId));
```

File: src/main/java/org/gridsuite/explore/server/services/FilterService.java [hunk 9]
```diff
@@ -79,13 +79,15 @@ public void insertFilter(String filter, UUID filterId, String userId) {
         restTemplate.exchange(filterServerBaseUri + path, HttpMethod.POST, httpEntity, Void.class);
     }
 
-    public void insertFilter(UUID sourceFilterId, UUID filterId, String userId) {
+    public UUID duplicateFilter(UUID filterId) {
         String path = UriComponentsBuilder.fromPath(DELIMITER + FILTER_SERVER_API_VERSION + "/filters")
-                .queryParam("duplicateFrom", sourceFilterId)
-                .queryParam("id", filterId)
+                .queryParam("duplicateFrom", filterId)
+                .buildAndExpand(filterId)
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/FilterService.java:85]: to remove

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 10]
```diff
@@ -72,6 +72,7 @@ public class ExploreTest {
     private static final UUID INVALID_ELEMENT_UUID = UUID.randomUUID();
     private static final UUID PARAMETERS_UUID = UUID.randomUUID();
     private static final UUID MODIFICATION_UUID = UUID.randomUUID();
+    private static final UUID NEW_ELEMENT_UUID = UUID.randomUUID();
```
Comment (sBouzols) [inline src/test/java/org/gridsuite/explore/server/ExploreTest.java:75]:
```suggestion
    private static final UUID PUBLIC_STUDY_COPY_UUID = UUID.randomUUID();
```

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 11]
```diff
@@ -151,13 +152,19 @@ public MockResponse dispatch(RecordedRequest request) {
...
+                            .addHeader("Content-Type", "application/json; charset=utf-8");
                 } else if (path.matches("/v1/studies.*") && "POST".equals(request.getMethod())) {
                     String bodyStr = body.readUtf8();
                     if (bodyStr.contains("filename=\"" + TEST_FILE_WITH_ERRORS + "\"")) {  // import file with errors
                         return new MockResponse().setResponseCode(409);
                     } else {
                         return new MockResponse().setResponseCode(200);
                     }
+                } else if (path.matches("/v1/cases\\?duplicateFrom=" + CASE_UUID + ".*") && "POST".equals(request.getMethod())) {
+                    return new MockResponse().setBody(newElementUuidAsString).setResponseCode(200)
```
Comment (sBouzols) [inline src/test/java/org/gridsuite/explore/server/ExploreTest.java:166]: then add a `CASE_COPY_UUID`

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 12]
```diff
@@ -221,9 +228,15 @@ public MockResponse dispatch(RecordedRequest request) {
                 } else if (path.matches("/v1/elements/.*") && "PUT".equals(request.getMethod())) {
                     return new MockResponse().setResponseCode(200)
                             .addHeader("Content-Type", "application/json; charset=utf-8");
+                } else if (path.matches("/v1/elements\\?duplicateFrom=.*&newElementUuid=.*") && "POST".equals(request.getMethod())) {
+                    return new MockResponse().setResponseCode(200)
```
Comment (sBouzols) [inline src/test/java/org/gridsuite/explore/server/ExploreTest.java:232]: no Body in response ?

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 13]
```diff
@@ -221,9 +228,15 @@ public MockResponse dispatch(RecordedRequest request) {
...
+                            .addHeader("Content-Type", "application/json; charset=utf-8");
                 } else if (path.matches("/v1/contingency-lists/metadata[?]ids=" + CONTINGENCY_LIST_UUID) && "GET".equals(request.getMethod())) {
                     return new MockResponse().setBody(listOfFormContingencyListAttributesAsString.replace("elementUuid", "id")).setResponseCode(200)
                             .addHeader("Content-Type", "application/json; charset=utf-8");
+                } else if (path.matches("/v1/.*contingency-lists\\?duplicateFrom=" + CONTINGENCY_LIST_UUID) && "POST".equals(request.getMethod())) {
+                    return new MockResponse().setBody(newElementUuidAsString).setResponseCode(200)
```
Comment (sBouzols) [inline src/test/java/org/gridsuite/explore/server/ExploreTest.java:238]: you cannot use the same UUID for different types I think, even for TUs it's confusing.

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java [hunk 14]
```diff
@@ -250,9 +266,12 @@ public MockResponse dispatch(RecordedRequest request) {
...
                 } else if (path.matches("/v1/identifier-contingency-lists/.*") && "PUT".equals(request.getMethod())) {
                     return new MockResponse().setResponseCode(200);
+                } else if (path.matches("/v1/parameters/" + PARAMETERS_UUID) && "POST".equals(request.getMethod())) {
+                    return new MockResponse().setBody(newElementUuidAsString).setResponseCode(200)
+                            .addHeader("Content-Type", "application/json; charset=utf-8");
                 } else if (path.matches("/v1/parameters.*")) {
                     return new MockResponse().setResponseCode(200);
-                } else if (path.matches("/v1/network-modifications/duplicate")) {
+                } else if (path.matches("/v1/network-modifications")) {
                     return new MockResponse().setBody(modificationIdsAsString).setResponseCode(200).addHeader("Content-Type", "application/json; charset=utf-8");
```
Comment (sBouzols) [inline src/test/java/org/gridsuite/explore/server/ExploreTest.java:275]: How could we expect a duplication mock endpoint return the same UUID as the duplicated entity, this is madness ! :D

File: src/main/java/org/gridsuite/explore/server/ExploreController.java [hunk 15]
```diff
@@ -177,19 +162,7 @@ public ResponseEntity<Void> createIdentifierContingencyList(@PathVariable("listN
...
-                                                          @RequestParam("listName") String listName,
-                                                          @RequestParam("description") String description,
-                                                          @RequestParam(QUERY_PARAM_PARENT_DIRECTORY_ID) UUID parentDirectoryUuid,
-                                                          @RequestHeader("userId") String userId) {
-        exploreService.duplicateIdentifierContingencyList(parentListId, listName, description, userId, parentDirectoryUuid);
-        return ResponseEntity.ok().build();
-    }
-
-    @PostMapping(value = "/explore/filters", consumes = MediaType.APPLICATION_JSON_VALUE)
+    @PostMapping(value = "/explore/filters", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/ExploreController.java:165]:
```suggestion
    @PostMapping(value = "/explore/filters", consumes = MediaType.APPLICATION_JSON_VALUE)
```

File: src/main/java/org/gridsuite/explore/server/ExploreController.java [hunk 16]
```diff
@@ -300,23 +272,21 @@ public ResponseEntity<Void> createParameters(@RequestBody String parameters,
...
-                                                 @RequestParam("name") String parametersName,
-                                                 @RequestParam(name = QUERY_PARAM_PARAMETERS_TYPE, defaultValue = "") ParametersType parametersType,
-                                                 @RequestParam(name = "description", required = false, defaultValue = "") String description,
-                                                 @RequestParam(QUERY_PARAM_PARENT_DIRECTORY_ID) UUID parentDirectoryUuid,
-                                                 @RequestHeader("userId") String userId) {
-        exploreService.duplicateParameters(parentParameterId, parametersType, parametersName, description, parentDirectoryUuid, userId);
+    public ResponseEntity<Void> duplicateParameters(@RequestParam("duplicateFrom") UUID parametersId,
+                                                    @RequestParam(name = QUERY_PARAM_PARENT_DIRECTORY_ID, required = false) UUID targetDirectoryId,
+                                                    @RequestParam(name = QUERY_PARAM_TYPE, defaultValue = "") ParametersType parametersType,
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/ExploreController.java:287]: why an empty defaultValue ?
> Supplying a default value implicitly sets required to false
Which service is called then ? is there a TU to test this ?

File: src/main/java/org/gridsuite/explore/server/services/ParametersService.java [hunk 17]
```diff
@@ -89,9 +89,8 @@ public void updateParameters(UUID parametersUuid, String parameters, ParametersT
     public UUID createParameters(UUID sourceParametersUuid, ParametersType parametersType) {
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/ParametersService.java:89]:
```suggestion
    public UUID duplicateParameters(UUID sourceParametersUuid, ParametersType parametersType) {
```

File: src/main/java/org/gridsuite/explore/server/services/ParametersService.java [hunk 18]
```diff
@@ -89,9 +89,8 @@ public void updateParameters(UUID parametersUuid, String parameters, ParametersT
...
         String parametersServerBaseUri = remoteServicesProperties.getServiceUri(genericParametersServices.get(parametersType));
         Objects.requireNonNull(sourceParametersUuid);
-        String urlParams = parametersType == ParametersType.VOLTAGE_INIT_PARAMETERS ? "?" + HEADER_DUPLICATE_FROM + "=" + sourceParametersUuid : "/" + sourceParametersUuid;
         var path = UriComponentsBuilder
-                    .fromPath(DELIMITER + SERVER_API_VERSION + "/parameters" + urlParams)
+                    .fromPath(DELIMITER + SERVER_API_VERSION + "/parameters/" + sourceParametersUuid)
```
Comment (sBouzols) [inline src/main/java/org/gridsuite/explore/server/services/ParametersService.java:93]: I see that if we change here with `duplicateFrom` form, we will need other PRs for each computaiton services... let's do it ?

File: src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java (hunk 1 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java (hunk 2 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java (hunk 3 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/services/ContingencyListService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java (hunk 5 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/services/ExploreService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java (hunk 6 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/services/DirectoryService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/DirectoryService.java (hunk 7 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/services/DirectoryService.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ExploreService.java (hunk 8 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/services/ExploreService.java]: done

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java (hunk 10 above)
Comment (anistouri) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java]: done

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java (hunk 11 above)
Comment (anistouri) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java]: done

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java (hunk 12 above)
Comment (anistouri) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java:243]: done

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java (hunk 13 above)
Comment (anistouri) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java]: done

File: src/test/java/org/gridsuite/explore/server/ExploreTest.java (hunk 14 above)
Comment (anistouri) [reply src/test/java/org/gridsuite/explore/server/ExploreTest.java:286]: yes, fixed

File: src/main/java/org/gridsuite/explore/server/ExploreController.java (hunk 15 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/ExploreController.java]: done

File: src/main/java/org/gridsuite/explore/server/ExploreController.java (hunk 16 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/ExploreController.java]: done

File: src/main/java/org/gridsuite/explore/server/services/ParametersService.java (hunk 18 above)
Comment (anistouri) [reply src/main/java/org/gridsuite/explore/server/services/ParametersService.java]: other PRs have been created

**Review by sBouzols — APPROVED** [review]
Code Review OK
Tests OK

**Pull request merged**